app/controllers/v1/tasks_controller.rb

```diff
@@ -0,0 +1,116 @@
+    end
+    def index
+      tasks, errors = ListTasks.new(index_params).call
+      if errors.any?
+        render json: { errors: errors }, status: 400
+      else
+        render json: tasks
+      end
+    end
+
+    swagger_api :show do
+      summary 'Fetch a single Task'
+      param :path, :project_id, :string, :required, 'Project Id'
+      param :path, :id, :string, :required, 'Task Id'
+    end
+    def show
+      task = Task.find_by id: params[:id]
+      if task.present?
+        render json: task
+      else
+        render json: { errors: ['Task not found'] }, status: 404
+      end
+    end
+
+    swagger_api :create do
+      summary 'Creates a new Task or Tasks'
+      param :path, :project_id, :string, :required, 'Project Id'
+      param :form, :name, :string, :optional, 'Task name'
+      param :form, :description, :string, :optional, 'Task description'
+      param :form, :state, :string, :optional, 'Task state'
+    end
+    def create
+      project = Project.find_by id: params[:project_id]
+      task = project.present? ? project.tasks.create(task_params) : nil
+      if project.present? && task.present?
+        render json: task, status: 201
+      else
+        render json: { errors: ['Unable to create task'] }, status: 400
+      end
+    end
+
+    swagger_api :update do
+      summary 'Updates an existing Task or Tasks'
+      param :path, :project_id, :string, :required, 'Project Id'
+      param :path, :id, :string, :required, 'Task Id'
+      param :form, :name, :string, :optional, 'Task name'
+      param :form, :description, :string, :optional, 'Task description'
+      param :form, :state, :string, :optional, 'Task state'
+    end
+
+    def update
+      task = Task.find_by(project_id: params[:project_id], id: params[:id])
+      if task.present? && task.update_attributes(task_params)
+        render json: task
+      elsif task.present?
+        render json: { errors: task.errors.full_messages }, status: 400
+      else
+        render json: { errors: ['Task not found'] }, status: 404
+      end
+    end
+
+    swagger_api :bulk_update do
+      summary 'Updates an existing Task or Tasks'
+      param :path, :project_id, :string, :required, 'Project Id'
+      param :path, :id, :string, :required, 'Task Id'
+      param :form, :project, :tasks_attributes, :required, 'Array of task attributes for nested update'
+    end
+
+    swagger_model :tasks_attributes do
+      property :tasks_attributes, :array, :required, 'Array of task attributes', { 'items' => { '$ref' => 'task' } }
+    end
+
+    swagger_model :task do
+      description 'nested task attributes for a specific project'
+      property :id, :string, :required, 'Task Id'
+      property :name, :string, :optional, 'Task name'
+      property :description, :string, :optional, 'Task description'
+      property :state, :string, :optional, 'Task state'
+    end
+
+    def bulk_update
```

Comment:
I might call this `bulk_upsert` as I believe it will insert or update the tasks for the project.

Reply:
Interesting @alexpeachey, I got curious on this one so I tested it out with Postman with the the following body (trying by omitting an id):

```
{ "project": {
   "tasks_attributes": [
        {
    
            "name": "X-Task 1",
            "description": "X-Task 1",
            "state": "todo"
        }]
}
}
```

I got

```
{
    "errors": [
        "Task not found"
    ]
}
```

with it as a "PUT" resource.  Did you mean I had to do something else to make it function as a bulk_upsert?  Just curious if it was something obvious I was missing, no need to answer.

Reply:
Interesting. `accepts_nested_attributes_for` is supposed to insert new associated records if you don't supply an id, and if you do supply an id it will update the existing record(s) assuming they exist in the collection.
What's even more interesting is your response:
"Task not found"
I'm fairly certain that is not a message you would get from `project.errors.full_messages` and the only other error in this controller action is `Project(s) not found`
You do show that error in `update` and `show`. So it's possible your message got routed to one of those??